core/api/serializers/business_plan.py

```diff
@@ -226,52 +238,93 @@ def get_agency(self, obj):
 
 
 class BPActivityCreateSerializer(serializers.ModelSerializer):
-    business_plan_id = serializers.PrimaryKeyRelatedField(
-        required=False,
-        queryset=BusinessPlan.objects.all().values_list("id", flat=True),
-    )
-    country_id = serializers.PrimaryKeyRelatedField(
-        queryset=Country.objects.all().values_list("id", flat=True),
-    )
+    # don't use `PrimaryKeyRelatedField`; makes 1 query / item when `many=True`
+    # FKs will be manually validated
+    business_plan_id = serializers.IntegerField(required=False)
+    country_id = serializers.IntegerField()
     lvc_status = serializers.ChoiceField(choices=BPActivity.LVCStatus.choices)
     project_type_id = serializers.IntegerField()
+    project_type_code = serializers.CharField(write_only=True)
     status = serializers.ChoiceField(choices=BPActivity.Status.choices)
-    bp_chemical_type_id = serializers.PrimaryKeyRelatedField(
-        queryset=BPChemicalType.objects.all().values_list("id", flat=True),
-    )
-    project_cluster_id = serializers.PrimaryKeyRelatedField(
-        queryset=ProjectCluster.objects.all().values_list("id", flat=True),
-    )
+    bp_chemical_type_id = serializers.IntegerField()
+    project_cluster_id = serializers.IntegerField()
 
-    substances = serializers.PrimaryKeyRelatedField(
-        many=True,
-        queryset=Substance.objects.all().values_list("id", flat=True),
-    )
-    comment_types = serializers.PrimaryKeyRelatedField(
-        required=False,
-        many=True,
-        queryset=CommentType.objects.all().values_list("id", flat=True),
-    )
+    # Many2Many represented as list of integers and manually validated
+    substances = Many2ManyListField(child=serializers.IntegerField())
+    comment_types = Many2ManyListField(child=serializers.IntegerField(), required=False)
 
     sector_id = serializers.IntegerField()
-    subsector_id = serializers.PrimaryKeyRelatedField(
-        queryset=ProjectSubSector.objects.all().values_list("id", flat=True),
-    )
+    sector_code = serializers.CharField(write_only=True)
+    subsector_id = serializers.IntegerField()
     values = BPActivityValueSerializer(many=True)
 
     is_updated = serializers.BooleanField(read_only=True)
 
-    def validate(self, attrs):
-        # check only once if project sector and type exist
-        sector = get_object_or_404(ProjectSector, id=attrs.get("sector_id"))
-        project_type = get_object_or_404(ProjectType, id=attrs.get("project_type_id"))
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        # get all related obj IDs only once for validation
```

Comment:
Based on the create method, the serializer creates a child instance for each object in the list. As a result, this optimization might not be necessary, since the initialization will be performed for each object, causing the related queries to be executed for every item individually. Or am I missing something?